src/transformers.jl

```diff
@@ -0,0 +1,66 @@
+
+"""
+    Transformation
+
+Abstract supertype for all transformations.
+"""
+abstract type Transformation end
+
+# Make Transforms callable types
+(t::Transformation)(x; kwargs...) = transform(x, t; kwargs...)
+
+"""
+    transform!(data::T, transformation::Transformation; kwargs...) -> T
+
+Apply `transformation` to the `data` in-place.
+Where possible, this should be extended for new data types `T`.
+"""
+function transform! end
+
+"""
+    transform(data::T, transformation::Transformation; kwargs...) -> T
+
+Non-in-place version of [`transform!`](@ref), which it delegates to by default.
+Does not need to be extended unless in-place transformation is not possible.
+"""
+function transform end
+
+"""
+    transform!(A::AbstractArray{T}, ::Transformation; dims=:, kwargs...) where T <: Real
+
+Applies the transformation to each element of `A`.
+Optionally specify the `dims` to apply the transformation along certain dimensions.
+"""
+function transform!(
+    A::AbstractArray{T}, t::Transformation; dims=:, kwargs...
```

Comment:
I'm picking a line at random here, but my comment is more general.

The present design requires that `kwargs` are propagated through the entire pipeline. This both creates clutter, and means that special-casing is required for things like `Colon()`.

We had a similar problem in [`KernelFunctions.jl`](https://github.com/JuliaGaussianProcesses/KernelFunctions.jl) where the initial design assumed that all inputs to functions were matrices, and you have to propagate an `obsdim` argument everywhere to say whether matrices we to be treated as "N x D" or "D x N" (D=dims, N=number of observations). The solution was to wrap the data itself in a thing that declared which way it was treated. This remove all kwargs and cleaned up the internals considerably.

A similar thing could be done here, specifically by
1. `Transformation`s do whatever transformation they encode to the entire object by default,
2. create `Transformation`s that apply to specific bits of data by composition. For example, introduce an `ApplyTo` `Transformation` `a` that eats another `Transformation` `t` and a "location specifier" -- calling `transform!` on `a` and data `x` would first pull out the bit of data in `x` that `a`'s "location specifier" refers to, and passes it off to `t`. `t` can then apply its transformation to the entirety of the data that it sees.

This would allow you to construct quite complicated transformations by composition e.g. specify that a transformation should only act on certain elements of a certain column by composing two `ApplyTo` transformations or something.

Is this a design choice that you would consider?

Reply:
Ohhhh, sorry, I had misunderstood how the present design works. I see what the `transform!` vs `_transform!` are doing now.

IIUC all stuff to do with "which bit of the data am I applying this to" lives in `transform!`, leaving `_transform!` free to operate on all of the data that is passed to it.

I'm still a little confused by the need to propagate kwargs everywhere other than the top layer though. Clearly you need to specify the dimensions to which a thing applies, so `transform!` needs them, but it's not clear to me that you need them in `_transform!` -- I would naively have imagined that the `Transformation` itself would contain all of the information needed to to apply itself to data. Is there a good example of where this isn't the case?

Reply:
> I would naively have imagined that the Transformation itself would contain all of the information needed to to apply itself to data
Just noting I also expected this.

Reply:
> The way I see it is, Transform -> transform is equivalent to ModelFactory -> Model in that sense.

Oh interesting -- this wasn't obvious to me.

As I understand it, the main reason that we make a distinction between `ModelFactory` and `Model` (and stuff along those lines generally) is because we want to apply the same `ModelFactory` to multiple distinct data sets, the details of which aren't known when the `ModelFactory` is constructed. We tend to build model factories outside of the context in which they'll get used, and combine them with problem-specific data when it becomes available (data set dimensions etc).

Conversely, I wasn't imagining that a `Transform` would generally get used in that way. In particular, unless I'm mistaken, we have the data to which a given `Transform` will be applied at construction time because we'll be building them inside `Forecasters.transform`. Or am I mistaken?

Reply:
So we could do that, but I was thinking that we might want to declaratively specify the transforms in the type. The default Forecaster.transform function would just run a default chain of transforms. This should make it more transparent to end users what operations are being applied, and ideally make it possible to modify that chain without rewriting the entire transform function.

Reply:
>In Impute.jl most of the methods aren't data specific and the dims keyword is pretty ubiquitous for all methods, so pushing that to the shared fallback methods made sense.

I'll mention that basing the design off of Impute was a deliberate choice, made for a few reasons
* Impute has gone through a few iterations in response to how it's been used - so we know it's been field tested
* We expect Transform and Impute to be interoperable, .e.g. compose chains of Imputors/Transforms

I'm open to tweaking the design but it would ideally be something that benefitted Impute as well.